File: raptor-backend-main/filter_avanti.js
/**
 * filter_uk.js
 * Extrait les opérateurs longue distance UK depuis le GTFS complet UK Rail.
 * Remplace filter_avanti.js (qui ne gardait que VT = Avanti).
 *
 * Agences conservées (national & intercités) :
 *
 *   — Intercités grande distance —
 *   VT  Avanti West Coast     Euston → Glasgow, Manchester, Birmingham, Liverpool
 *   GR  LNER                  Kings Cross → Edinburgh, Newcastle, Leeds, York
 *   CS  Caledonian Sleeper    Euston → Inverness, Aberdeen, Fort William (nuit)
 *   XC  CrossCountry          Birmingham → Edinburgh/Glasgow/Bristol/Plymouth
 *   TP  TransPennine Express  Manchester/Liverpool → Edinburgh, Glasgow, Newcastle
 *   EM  East Midlands Railway St Pancras → Nottingham, Sheffield, Leeds, Derby
 *   GW  Great Western Railway Paddington → Bristol, Cardiff, Plymouth, Cornwall
 *   SW  South Western Railway Waterloo → Southampton, Bournemouth, Exeter
 *   HT  Hull Trains           Kings Cross → Hull (open access)
 *   GC  Grand Central         Kings Cross → Sunderland, Bradford (open access)
 *   LD  Lumo                  Kings Cross → Edinburgh (low-cost open access)
 *
 *   — Réseaux nationaux régionaux —
 *   SR  ScotRail              Réseau national écossais (Glasgow, Edinburgh, Inverness…)
 *   NT  Northern Trains       Nord Angleterre (Manchester, Leeds, Newcastle, Liverpool…)
 *   AW  Transport for Wales   Réseau national gallois (Cardiff, Swansea, Holyhead…)
 *
 *   — Exclus (urbain/banlieue) —
 *   LT London Underground · XR Elizabeth line · LO London Overground (métro londonien)
 *   TL Thameslink · GN Great Northern · CC c2c · SN Southern · SE Southeastern (banlieue)
 *   ME Merseyrail (métro Liverpool) · HX Heathrow Express · GX Gatwick Express · IL Island Line
 *
 * Source : GTFS UK Rail complet (transit.land f-uk~rail)
 * Sortie  : ./gtfs/UK_Intercity/
 */

const fs       = require('fs');
const path     = require('path');
const readline = require('readline');

const SOURCE_DIR = './gtfs/UK_Rail';
const TARGET_DIR = './gtfs/UK_National';

// Agences nationales conservées (intercités + réseaux nationaux régionaux)
const TARGET_AGENCIES = new Set([
  // ── Intercités grande distance ──────────────────────────────────────────
  'VT',  // Avanti West Coast     — Euston → Glasgow, Manchester, Birmingham, Liverpool
  'GR',  // LNER                  — Kings Cross → Edinburgh, Newcastle, Leeds, York
  'CS',  // Caledonian Sleeper    — Euston → Inverness, Aberdeen, Fort William (nuit)
  'XC',  // CrossCountry          — Birmingham → Edinburgh/Glasgow/Bristol/Plymouth
  'TP',  // TransPennine Express  — Manchester/Liverpool → Edinburgh, Glasgow, Newcastle
  'EM',  // East Midlands Railway — St Pancras → Nottingham, Sheffield, Leeds, Derby
  'GW',  // Great Western Railway — Paddington → Bristol, Cardiff, Plymouth, Cornwall
  'SW',  // South Western Railway — Waterloo → Southampton, Bournemouth, Exeter
  'HT',  // Hull Trains           — Kings Cross → Hull (open access)
  'GC',  // Grand Central         — Kings Cross → Sunderland, Bradford (open access)
  'LD',  // Lumo                  — Kings Cross → Edinburgh (low-cost open access)
  // ── Réseaux nationaux régionaux ─────────────────────────────────────────
  'SR',  // ScotRail              — Réseau national écossais (Glasgow, Edinburgh, Inverness, Aberdeen…)
  'NT',  // Northern Trains       — Nord Angleterre (Manchester, Leeds, Newcastle, Liverpool, Hull…)
  'AW',  // Transport for Wales   — Réseau national gallois (Cardiff, Swansea, Holyhead…)
]);

if (!fs.existsSync(TARGET_DIR)) fs.mkdirSync(TARGET_DIR, { recursive: true });

// ── Lecture CSV streaming ──────────────────────────────────────────────────────
function parseCSVLine(line) {
  const result = []; let cur = ''; let inQ = false;
  for (const c of line) {
    if (c === '"')              { inQ = !inQ; }
    else if (c === ',' && !inQ) { result.push(cur); cur = ''; }
    else                        { cur += c; }
  }
  result.push(cur);
  return result;
}

function processCSV(srcFile, dstFile, onRow) {
  return new Promise((resolve, reject) => {
    if (!fs.existsSync(srcFile)) {
      console.warn(`  ⚠  Absent : ${path.basename(srcFile)}`);
      return resolve({ count: 0 });
    }
    const input  = fs.createReadStream(srcFile, { encoding: 'utf8' });
    const output = fs.createWriteStream(dstFile, { encoding: 'utf8' });
    const rl     = readline.createInterface({ input, crlfDelay: Infinity });
    let headers  = null;
    let count    = 0;
    rl.on('line', (raw) => {
      const line = raw.replace(/^\uFEFF/, '').trim();
      if (!line) return;
      if (!headers) { headers = parseCSVLine(line); output.write(line + '\n'); return; }
      const cols = parseCSVLine(line);
      const row  = {};
      headers.forEach((h, i) => { row[h] = (cols[i] ?? '').trim(); });
      if (onRow(row)) { output.write(line + '\n'); count++; }
    });
    rl.on('close', () => { output.end(() => resolve({ count })); });
    rl.on('error', reject);
    output.on('error', reject);
  });
}

// ── Pipeline ──────────────────────────────────────────────────────────────────
async function filterUK() {
  console.log(`\n⚙️  Filtrage UK national & intercités...`);
  console.log(`   Agences : ${[...TARGET_AGENCIES].join(', ')}`);
  console.log(`   Source  : ${SOURCE_DIR}`);
  console.log(`   Cible   : ${TARGET_DIR}\n`);

  // 1. agency.txt
  const { count: agencyCount } = await processCSV(
    `${SOURCE_DIR}/agency.txt`, `${TARGET_DIR}/agency.txt`,
    (row) => TARGET_AGENCIES.has(row.agency_id)
  );
  console.log(`  agency.txt       : ${agencyCount} agence(s)`);

  // 2. routes.txt → mémoriser route_ids
  const routeIds = new Set();
  const { count: routeCount } = await processCSV(
    `${SOURCE_DIR}/routes.txt`, `${TARGET_DIR}/routes.txt`,
    (row) => {
      if (!TARGET_AGENCIES.has(row.agency_id)) return false;
      routeIds.add(row.route_id);
      return true;
    }
  );
  console.log(`  routes.txt       : ${routeCount} route(s)`);

  // 3. trips.txt → mémoriser trip_ids + service_ids
  const tripIds    = new Set();
  const serviceIds = new Set();
  const { count: tripCount } = await processCSV(
    `${SOURCE_DIR}/trips.txt`, `${TARGET_DIR}/trips.txt`,
    (row) => {
      if (!routeIds.has(row.route_id)) return false;
      tripIds.add(row.trip_id);
      serviceIds.add(row.service_id);
      return true;
    }
  );
  console.log(`  trips.txt        : ${tripCount} trip(s)`);

  // 4. stop_times.txt → mémoriser stop_ids utilisés
  const stopIds = new Set();
  const { count: stCount } = await processCSV(
    `${SOURCE_DIR}/stop_times.txt`, `${TARGET_DIR}/stop_times.txt`,
    (row) => {
      if (!tripIds.has(row.trip_id)) return false;
      stopIds.add(row.stop_id);
      return true;
    }
  );
  console.log(`  stop_times.txt   : ${stCount} ligne(s)`);

  // 5. stops.txt — arrêts utilisés + stations parentes
  const { count: stopCount } = await processCSV(
    `${SOURCE_DIR}/stops.txt`, `${TARGET_DIR}/stops.txt`,
    (row) => stopIds.has(row.stop_id) || row.location_type === '1'
  );
  console.log(`  stops.txt        : ${stopCount} arrêt(s)`);

  // 6. calendar.txt
  if (fs.existsSync(`${SOURCE_DIR}/calendar.txt`)) {
    const { count: calCount } = await processCSV(
      `${SOURCE_DIR}/calendar.txt`, `${TARGET_DIR}/calendar.txt`,
      (row) => serviceIds.has(row.service_id)
    );
    console.log(`  calendar.txt     : ${calCount} service(s)`);
  }

  // 7. calendar_dates.txt
  if (fs.existsSync(`${SOURCE_DIR}/calendar_dates.txt`)) {
    const { count: cdCount } = await processCSV(
      `${SOURCE_DIR}/calendar_dates.txt`, `${TARGET_DIR}/calendar_dates.txt`,
      (row) => serviceIds.has(row.service_id)
    );
    console.log(`  calendar_dates   : ${cdCount} exception(s)`);
  }

  // 8. feed_info.txt
  if (fs.existsSync(`${SOURCE_DIR}/feed_info.txt`)) {
    fs.copyFileSync(`${SOURCE_DIR}/feed_info.txt`, `${TARGET_DIR}/feed_info.txt`);
    console.log(`  feed_info.txt    : copié`);
  }

  console.log(`\n✅ UK national filtré → ${TARGET_DIR}`);
  console.log(`   ${routeIds.size} routes · ${tripIds.size} trips · ${stopIds.size} arrêts\n`);
}

filterUK().catch(err => {
  console.error('❌ Erreur filtrage UK :', err);
  process.exit(1);
});

File: raptor-backend-main/filter_avanti.py
import os
import csv

# Configuration
SOURCE_DIR = "./gtfs/UK_Rail"
TARGET_DIR = "./gtfs/Avanti_Only"
TARGET_AGENCY_ID = "VT"

if not os.path.exists(TARGET_DIR):
    os.makedirs(TARGET_DIR)

def filter_gtfs():
    print(f"--- Début de l'extraction pour l'agence : {TARGET_AGENCY_ID} ---")

    # 1. Filtrer AGENCY.TXT
    print("Filtrage de agency.txt...")
    with open(f"{SOURCE_DIR}/agency.txt", 'r', encoding='utf-8') as f_in, \
         open(f"{TARGET_DIR}/agency.txt", 'w', encoding='utf-8', newline='') as f_out:
        reader = csv.DictReader(f_in)
        writer = csv.DictWriter(f_out, fieldnames=reader.fieldnames)
        writer.writeheader()
        for row in reader:
            if row['agency_id'] == TARGET_AGENCY_ID:
                writer.writerow(row)

    # 2. Filtrer ROUTES.TXT et mémoriser les route_ids
    print("Filtrage de routes.txt...")
    route_ids = set()
    with open(f"{SOURCE_DIR}/routes.txt", 'r', encoding='utf-8') as f_in, \
         open(f"{TARGET_DIR}/routes.txt", 'w', encoding='utf-8', newline='') as f_out:
        reader = csv.DictReader(f_in)
        writer = csv.DictWriter(f_out, fieldnames=reader.fieldnames)
        writer.writeheader()
        for row in reader:
            if row['agency_id'] == TARGET_AGENCY_ID:
                writer.writerow(row)
                route_ids.add(row['route_id'])

    # 3. Filtrer TRIPS.TXT et mémoriser les trip_ids et service_ids
    print("Filtrage de trips.txt...")
    trip_ids = set()
    service_ids = set()
    with open(f"{SOURCE_DIR}/trips.txt", 'r', encoding='utf-8') as f_in, \
         open(f"{TARGET_DIR}/trips.txt", 'w', encoding='utf-8', newline='') as f_out:
        reader = csv.DictReader(f_in)
        writer = csv.DictWriter(f_out, fieldnames=reader.fieldnames)
        writer.writeheader()
        for row in reader:
            if row['route_id'] in route_ids:
                writer.writerow(row)
                trip_ids.add(row['trip_id'])
                service_ids.add(row['service_id'])

    # 4. Filtrer STOP_TIMES.TXT (Le plus lourd, traité ligne par ligne)
    print("Filtrage de stop_times.txt (cela peut prendre du temps)...")
    stop_ids = set()
    with open(f"{SOURCE_DIR}/stop_times.txt", 'r', encoding='utf-8') as f_in, \
         open(f"{TARGET_DIR}/stop_times.txt", 'w', encoding='utf-8', newline='') as f_out:
        reader = csv.DictReader(f_in)
        writer = csv.DictWriter(f_out, fieldnames=reader.fieldnames)
        writer.writeheader()
        for row in reader:
            if row['trip_id'] in trip_ids:
                writer.writerow(row)
                stop_ids.add(row['stop_id'])

    # 5. Filtrer STOPS.TXT
    print("Filtrage de stops.txt...")
    with open(f"{SOURCE_DIR}/stops.txt", 'r', encoding='utf-8') as f_in, \
         open(f"{TARGET_DIR}/stops.txt", 'w', encoding='utf-8', newline='') as f_out:
        reader = csv.DictReader(f_in)
        writer = csv.DictWriter(f_out, fieldnames=reader.fieldnames)
        writer.writeheader()
        for row in reader:
            # On garde l'arrêt s'il est utilisé par Avanti ou si c'est une station parente
            if row['stop_id'] in stop_ids or row['location_type'] == '1':
                writer.writerow(row)

    # 6. Filtrer CALENDAR.TXT
    print("Filtrage de calendar.txt...")
    if os.path.exists(f"{SOURCE_DIR}/calendar.txt"):
        with open(f"{SOURCE_DIR}/calendar.txt", 'r', encoding='utf-8') as f_in, \
             open(f"{TARGET_DIR}/calendar.txt", 'w', encoding='utf-8', newline='') as f_out:
            reader = csv.DictReader(f_in)
            writer = csv.DictWriter(f_out, fieldnames=reader.fieldnames)
            writer.writeheader()
            for row in reader:
                if row['service_id'] in service_ids:
                    writer.writerow(row)

    # 7. Copier FEED_INFO.TXT (Optionnel)
    if os.path.exists(f"{SOURCE_DIR}/feed_info.txt"):
        import shutil
        shutil.copyfile(f"{SOURCE_DIR}/feed_info.txt", f"{TARGET_DIR}/feed_info.txt")

    print(f"--- Terminé ! Les données sont dans {TARGET_DIR} ---")

if __name__ == "__main__":
    filter_gtfs()